### ISSUE TITLE
The code of the added mruby library does not work

### ISSUE THREAD

**Issue by makiyamaju** [h2o/h2o#1356]
Hello
I am trying to add a mruby library to h2o and run it.
I added a library to deps under h2o and wrote the code in mruby's handler file but it does not work.
For example, if you add mruby-os and run the code OS.machine, it becomes Internal Server Error. It works on mirb.
Is there anything else to do except put the library in deps?
The environment is Virtualbox + centos 7.3.

**Comment by yannick** [comment]
hi @makiyamaju 
it depends on whether its just mruby code or mruby with c code.
for the later (so also for mruby-os)  you need to add it when you build h2o, i just documented this in https://github.com/h2o/h2o/pull/1323 which is probably not yet online.

if its ruby source code only you should be able to load it at runtime

**Comment by makiyamaju** [comment]
Mr. yannick
Thank you for your reply.

I read # 1323, but I could not understand it well.

I understood as follows.

In the case of pure ruby libraly.

1 Place the library in git in the deps directory.
2 cmake
3 make
4 make install
3 The library is loaded.

For ruby + c language library.

1 Place the library in git in the deps directory.
2 cmake - ooooooo (add something)
3 make
4 make install
3 The library is loaded.

> For the later (so also for mruby-os) you need to add it

What should I add to what?

I looked up to mrb_vm_exec using GDB.
However, I did not know the cause of Internal Server Error,

Thank you.

Makiyamajun

**Comment by yannick** [comment]
hi @makiyamaju 

i just tested this on macOS and it works:
note: on linux you can use normal readlink and possibly omit `-DOPENSSL_ROOT_DIR` 
```
git clone https://github.com/h2o/h2o.git h2o-test-mruby-os
cd h2o-test-mruby-os

git submodule add https://github.com/appPlant/mruby-os.git deps/mruby-os

mkdir build
mkdir install
cd build
cmake -DWITH_BUNDLED_SSL=off -DWITH_MRUBY=ON -LAHT -DCMAKE_BUILD_TYPE=Debug -DOPENSSL_ROOT_DIR=/usr/local/Cellar/openssl@1.1/1.1.0f -DCMAKE_INSTALL_PREFIX=$(greadlink -f ../install) ..

make h2o #no need to build libh2o
make install #this is important so the .rb files are being found!

curl 'https://gist.githubusercontent.com/yannick/337797c1b1a14ff339cb5ae819187038/raw/ca44e783c96c9eb9952dd7b31e3d60d190d802ba/h2o.conf' -o ../install/h2o.conf

cd ../install
./bin/h2o -c h2o.conf
```

then:

```
curl localhost:8080/ -i
HTTP/1.1 200 OK
Date: Fri, 14 Jul 2017 09:31:42 GMT
Connection: keep-alive
Content-Length: 38
Server: h2o/2.3.0-DEV
content-type: text/plain

hello world, i'm running h2o on Darwin%
```

**Comment by makiyamajun** [comment]
Mr. yannick

Thank you for reply

I improved my procedure and h2oconf with reference to your reply.
My environment is virtualbox + centos 7.3, so I modified your procedure and used it.

The way to get my library, moved to deps and typed the following.
Git clone https://github.com/mattn/mruby-md5.git
I think it is basically the same, but the situation got better.

And I used make h2o instead of make.

From the home directory.

git clone https://github.com/h2o/h2o.git
cd h2o
git submodule update --recursive

git submodule add https://github.com/appPlant/mruby-os.git deps/mruby-os
git submodule add https://github.com/Asmod4n/mruby-b64.git deps/mruby-b64
git submodule add https://github.com/Asmod4n/mruby-b64.git deps/mruby-base64
git submodule add https://github.com/mattn/mruby-md5.git deps/mruby-md5
git submodule add https://github.com/matsumotory/mruby-sleep deps/mruby-sleep
git submodule add https://github.com/ksss/mruby-stringio deps/mruby-stringio

cmake -DWITH_BUNDLED_SSL=off -DWITH_MRUBY=ON -DCMAKE_BUILD_TYPE=Debug .
make h2o
make install
cd ../

/usr/local/bin/h2o --mode=master -c h2o.conf

And I added some library behavior to h2o.conf.
The result is as follows.
There are libraries that do not work and do.
The message of the library that does not work is Internal Server Error.

Can I run these libraries?

h2o.conf

listen: 80

hosts:
  "127.0.0.1:80":
    paths:
      /:
        mruby.handler: |

            #html = "Hello mruby in h2o"                              # text OK
            #html = "hello world, i'm running h2o on #{OS.sysname}"  # mruby-os      OK
            #html = "#{Base64::encode('ruby')}"                       　　　　　　# mruby-base64 Internal Server Error
            #html = "#{MD5::md5_hex('ruby')}"                 　　　　　　　　　 # mruby-md5    OK                    
            "#{Sleep::sleep(10)}"                              　　　　　　　　　      　　 # mruby-sleep  OK
            html = "sleep 10s"                                 

            #io = StringIO.new                                         　　　　　　# mruby-stringio Internal Server Error
            #io.puts "Hello World"
            #html = io.string #=> "Hello World\n"

            Proc.new do |env|
            [200, {'content-type' => 'text/plain'}, [html]]
            end

    access-log: /dev/stdout


Thank you.

**Comment by yannick** [comment]
cool. i think there should be more detailed errors in your error log, maybe set `error-log: /path/to/error-log-file` explicitly.
for sleep note that there is built in sleep now: https://github.com/h2o/h2o/pull/1348

**Comment by makiyamajun** [comment]
Mr. yannick

Is sleep already there?
Thank you.
Since evolution is fast, h2o is hard to follow.

> i think there should be more detailed errors in your error log, 
that's right

The error log is as follows.
I hope I have more information.
I can not fix errors with this information.

starting new worker 26518
[INFO] raised RLIMIT_NOFILE to 4096
h2o server (pid:26518) is ready to serve requests
mruby raised: h2o2.conf:16:uninitialized constant StringIO (NameError)
127.0.0.1 - - [17/Jul/2017:19:05:35 +0900] "GET / HTTP/1.1" 500 21 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firef

Thank you.

makiyamajun

**Comment by yannick** [comment]
this points towards https://github.com/ksss/mruby-stringio/blob/master/mrblib/stringio.rb 
not being included properly. it should be in `/usr/share/h2o/mruby/`. 
you can copy it manually there for now. 

 @kazuho  shouldn't ruby files from mrblib automatically be copied or what is the process to get those into the mruby dir?

**Comment by kazuho** [comment]
> this points towards https://github.com/ksss/mruby-stringio/blob/master/mrblib/stringio.rb
not being included properly. it should be in `/usr/share/h2o/mruby/`.

No. The files under `deps/mruby-*/mrblib` are embedded as part of libmruby, and h2o links against the library.

If you have anything missing, please run a full rebuild and see what happens.
